### PR TITLE
client: protect against reading too much data from stdin

### PR DESCRIPTION
**Pull request by mvo5**
This is a followup for PR#9692 where it was suggested that the
new "snapctl" support for reading from stdin should have some
protection against reading too much data. See
https://github.com/snapcore/snapd/pull/9692#discussion_r530255566

This starts with a 4MB limit.

### CODE REVIEW

**Review by zyga — COMMENTED** [review]
This needs to be documented somewhere on `snapctl` forum pages (CC @degville)

**Review by anonymouse64 — APPROVED** [review]
LGTM, thanks for this cleanup

File: client/snapctl.go [hunk 1]
```diff
@@ -62,16 +62,23 @@ type snapctlOutput struct {
...
+var stdinReadLimit = int64(4 * 1000 * 1000)
+
 // RunSnapctl requests a snapctl run for the given options.
 func (client *Client) RunSnapctl(options *SnapCtlOptions, stdin io.Reader) (stdout, stderr []byte, err error) {
 	// TODO: instead of reading all of stdin here we need to forward it to
 	//       the daemon eventually
 	var stdinData []byte
 	if stdin != nil {
-		stdinData, err = ioutil.ReadAll(stdin)
+		limitedStdin := &io.LimitedReader{R: stdin, N: stdinReadLimit}
```
Comment (bboozzoo) [inline client/snapctl.go:74]: Unfortunately LimitedReader does not make it easy to detect when it hit the limit if one is not invoking Read() manually and inspecting N afterwards. So this should be stdinReadLimit + 1 to have a 4MB limit with the current logic AFAIU.

File: client/snapctl.go [hunk 2]
```diff
@@ -62,16 +62,23 @@ type snapctlOutput struct {
...
+		stdinData, err = ioutil.ReadAll(limitedStdin)
 		if err != nil {
 			return nil, nil, fmt.Errorf("cannot read stdin: %v", err)
 		}
+		if limitedStdin.N <= 0 {
+			return nil, nil, fmt.Errorf("cannot read more than %v bytes of data from stdin", stdinReadLimit-1)
```
Comment (bboozzoo) [inline client/snapctl.go:80]: And then here we can use `stdinReadLimit`

File: client/snapctl_test.go [hunk 3]
```diff
@@ -82,3 +82,24 @@ func (cs *clientSuite) TestInternalSnapctlCmdNeedsStdin(c *check.C) {
...
+
+func (cs *clientSuite) TestClientRunSnapctlReadLimit(c *check.C) {
+	cs.rsp = `{
+		"type": "sync",
+        "status-code": 200,
+		"result": {
+		}
+	}`
+
+	restore := client.MockStdinReadLimit(11)
```
Comment (bboozzoo) [inline client/snapctl_test.go:94]: and 10 here.

Comment (bboozzoo) [inline client/snapctl_test.go:94]: Also could use a test where we give it stdinReadLimit bytes in the input and so no error should be returned.

File: client/snapctl.go (hunk 1 above)
Comment (stolowski) [reply client/snapctl.go:74]: Hmm I'm not sure this is true unless I'm missing something... ioutil.ReadAll is a wrapper for io.ReadAll which wraps Read; LimitedReader's Read() is guaranteed to stop reading with EOF upon reaching N; N can be inspected afterwards.

https://golang.org/src/io/io.go?s=15540:15633#L451

Comment (bboozzoo) [reply client/snapctl.go:74]: When the input is exactly N bytes, N will be 0 too, and AFAICT it never goes < 0, which looks identical to a case when there was more bytes in the input and LimitedReader returned EOF early. See: https://play.golang.org/p/MPlRZmrx11s So you cannot really distinguish a case from a valid input of exactly N bytes, and too long one that was trimmed.

Comment (stolowski) [reply client/snapctl.go:74]: Ahhh, got you, thanks for the example, that's a bummer /o\

**Comment by pedronis** [comment]
>  This needs to be documented somewhere on snapctl forum pages (CC @degville)

@zyga, @degville  we actually haven't documented yet the functionality this is related to, because it not finalized yet

File: client/snapctl_test.go (hunk 3 above)
Comment (mvo5) [reply client/snapctl_test.go]: Added

**Review by bboozzoo — APPROVED** [review]
LGTM

**Pull request merged**